netshare/generate/denormalize_fields.py

```diff
@@ -67,72 +74,160 @@ def write_to_csv(
+                    )
+                    .transpose()
+                    .astype(float)
+                )
+                timestamp_matrix = np.expand_dims(
+                    np.add(flow_start_expand, interarrival_cumsum), axis=2
+                )
+                timeseries_numpy = np.concatenate(
+                    (timeseries_numpy, timestamp_matrix), axis=2
+                )
+                timeseries_titles_idx.append(timeseries_numpy.shape[2] - 1)
+
         writer.writerow(session_titles + timeseries_titles)
 
+        session_key_set = set()
         for (
             data_gen_per_session,
             session_data_per_session,
             timeseries_per_session,
-        ) in zip(data_gen_flag, session_key, timeseries):
-            # this if is here in parallel to the if in `reduce_samples`. It supports old flows.
+        ) in zip(
+            data_gen_flag,
+            session_key_numpy[:, session_titles_idx],  # remove cols not in raw data
+            timeseries_numpy[
+                :, :, timeseries_titles_idx
+            ],  # remove cols not in raw data
+        ):
             session_data_per_session = session_data_per_session.tolist()
+            # remove duplicated session keys
+            if tuple(session_data_per_session) in session_key_set:
+                logger.debug(f"Session key {session_data_per_session} already exists!")
+                continue
+            session_key_set.add(tuple(session_data_per_session))
             for j in range(data_gen_per_session.shape[0]):
                 if data_gen_per_session[j] == 1.0:
                     timeseries_data = timeseries_per_session[j].tolist()
                     writer.writerow(session_data_per_session + timeseries_data)
 
 
-def denormalize_fields() -> str:
+def denormalize_fields() -> None:
     """
-    This function denormalizes the data in the generated_data folder using the attributes
-        and features fields that were created in the pre-process step.
+    This function denormalizes the data in the generated_data folder using the attributes and features fields that were created in the pre-process step.
     Last, it writes the denormalized data to a csv file under the same directory hierarchy as the created data.
 
     :return: the path to the denormalized data.
     """
-    output_folder = tempfile.mkdtemp()
-
-    session_key_fields = list(learn_api.get_attributes_fields().values())
-    timeseries_fields = list(learn_api.get_feature_fields().values())
-
-    for (
-        unnormalized_timeseries,
-        unnormalized_session_key,
-        data_gen_flag,
-        sub_folder,
-        filename,
-    ) in generate_api.get_raw_generated_data():
-        session_key = _denormalize_by_fields_list(
-            unnormalized_session_key, session_key_fields, is_session_key=True
+    configs, config_group_list = create_chunks_configurations(generation_flag=True)
+
+    for config in tqdm(configs):
```

Comment:
do we want the `tqdm` here? does it take a lot of time?
if yeah, then check out the other places on how to do it - we want it to be logged, and that it will tell us "what are we waiting for" :)